pkg/platform/runtime/buildplan/buildplan.go

```diff
@@ -18,18 +18,41 @@ func NewMapFromBuildPlan(build *model.Build) (artifact.Map, error) {
 	lookup := make(map[strfmt.UUID]interface{})
 
 	for _, artifact := range build.Artifacts {
-		lookup[artifact.TargetID] = artifact
+		lookup[artifact.NodeID] = artifact
 	}
 	for _, step := range build.Steps {
-		lookup[step.TargetID] = step
+		lookup[step.StepID] = step
 	}
 	for _, source := range build.Sources {
-		lookup[source.TargetID] = source
+		lookup[source.NodeID] = source
 	}
 
 	var terminalTargetIDs []strfmt.UUID
 	for _, terminal := range build.Terminals {
-		terminalTargetIDs = append(terminalTargetIDs, terminal.TargetIDs...)
+		// If there is an artifact for this terminal and its mime type is application/x-gozip-installer
+		// then the list of terminal target IDs should be the sources for the step that generated the
+		// installer. That is what we want to use to build the artifact map.
+		for _, nodeID := range terminal.NodeIDs {
+			artifact, ok := lookup[nodeID].(*model.Artifact)
+			if !ok {
+				continue
+			}
+
+			if artifact.MimeType == "application/x-gozip-installer" {
```

Comment:
This feels overly opinionated on specifically the gozip installer. If it were up to me this should never be part of the client-side logic and we should be looking for something more generic. Can you check with Tyler/Martin if there is a more appropriate thing that we can check for to mean "look up artifacts from this terminal".

Reply:
Yeah, I agree.  If we add a different type of installer/packager like a docker container this won't work anymore.

My personal preference was that we have more specific terminals such that you don't need this graph walking at all.  So, the buildplan would have terminals with tags like this:

* `"platform:<UUID>:gozip-installer" == {"<UUID of gozip-artifact>"}`
* `"platform:<UUID>:state-tool-artifacts" == {"<UUID of artifact 1>", "<UUID of artifact 2>"}`

But I believe that Scott didn't like this for some reason.  @srobertson : Do you remember why you preferred that the clients would walk the graph here?

Assuming that there is a good reason, I would instead check for `artifact.MimeType != <State Tool Artifact Mime Type> || artifact.MimeType != <Camel Installer Mime Type>`, because these are the only ones the State Tool can handle.

So, basically you walk up the graph until you find what you are looking for oblivious of what it might be.